src/backend/distributed/planner/multi_physical_planner.c

```diff
@@ -1271,38 +1284,87 @@ JoinRangeTableEntry(JoinExpr *joinExpr, List *dependentJobList, List *rangeTable
 	rangeTableEntry->eref->colnames = joinedColumnNames;
 	rangeTableEntry->joinaliasvars = joinedColumnVars;
 
-	SetJoinRelatedColumnsCompat(rangeTableEntry,
-								leftColumnNames, rightColumnNames, leftColumnVars,
+	SetJoinRelatedColumnsCompat(rangeTableEntry, leftRelId, rightRelId, leftColumnVars,
 								rightColumnVars);
 
 	return rangeTableEntry;
 }
 
 
 static void
-SetJoinRelatedColumnsCompat(RangeTblEntry *rangeTableEntry,
-							List *leftColumnNames, List *rightColumnNames,
+SetJoinRelatedColumnsCompat(RangeTblEntry *rangeTableEntry, Oid leftRelId, Oid rightRelId,
 							List *leftColumnVars, List *rightColumnVars)
 {
 	#if PG_VERSION_NUM >= PG_VERSION_13
 
 	/* We don't have any merged columns so set it to 0 */
 	rangeTableEntry->joinmergedcols = 0;
-	int numvars = list_length(leftColumnVars);
-	for (int varId = 1; varId <= numvars; varId++)
+
+	if (OidIsValid(leftRelId))
 	{
-		rangeTableEntry->joinleftcols = lappend_int(rangeTableEntry->joinleftcols, varId);
+		rangeTableEntry->joinleftcols = GetColumnOriginalIndexes(leftRelId);
 	}
-	numvars = list_length(rightColumnVars);
-	for (int varId = 1; varId <= numvars; varId++)
+	else
+	{
+		rangeTableEntry->joinleftcols = CreateJoinColIndexes(leftColumnVars);
+	}
+
+	if (OidIsValid(rightRelId))
+	{
+		rangeTableEntry->joinrightcols = GetColumnOriginalIndexes(rightRelId);
+	}
+	else
 	{
-		rangeTableEntry->joinrightcols = lappend_int(rangeTableEntry->joinrightcols,
-													 varId);
+		rangeTableEntry->joinrightcols = CreateJoinColIndexes(rightColumnVars);
 	}
+
 	#endif
 }
 
 
+#if PG_VERSION_NUM >= PG_VERSION_13
+
+/*
+ * CreateJoinColIndexes creates join column indexes based on the given columnVars.
+ */
+static List *
+CreateJoinColIndexes(List *columnVars)
+{
+	List *joinColIndexes = NIL;
+	int numvars = list_length(columnVars);
+	for (int varId = 1; varId <= numvars; varId++)
+	{
+		joinColIndexes = lappend_int(joinColIndexes, varId);
+	}
```

Comment:
it might be better if we can define a listutils function (`CreateIntListFromSequence` ?) to generate an int list from a closed range and use it here, what do you think ?

Reply:
Sure, added the utility function